numba/cuda/tests/cudapy/test_cooperative_groups.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import print_function
+
+import numpy as np
+
+from numba import config, cuda, int32, void
+from numba.cuda.testing import unittest, CUDATestCase, skip_on_cudasim
+
+
+def cc_X_or_above(major, minor):
+    if not config.ENABLE_CUDASIM:
+        cc = cuda.current_context().device.compute_capability
+        return cc >= (major, minor)
+    else:
+        return True
+
+
+def skip_unless_cc_60(fn):
+    return unittest.skipUnless(cc_X_or_above(6, 0), "requires cc >= 6.0")(fn)
+
+
+@cuda.jit
+def this_grid(A):
+    cuda.cg.this_grid()
+    A[0] = 1.0
+
+
+@cuda.jit
+def sync_group(A):
+    g = cuda.cg.this_grid()
+    g.sync()
+    A[0] = 1.0
+
+
+@cuda.jit
+def no_sync(A):
+    A[0] = cuda.grid(1)
+
+
+def sequential_rows(M):
+    # The grid writes rows one at a time. Each thread reads an element from
+    # the previous row written by its "opposite" thread.
+    #
+    # A failure to sync the grid at each row would result in an incorrect
+    # result as some threads could run ahead of threads in other blocks, or
+    # fail to see the update to the previous row from their opposite thread.
```

Comment:
I'm not sure that this is a would, more of a could? If you don't sync the grid across rows does it always fail, just wondering if there's something hardware-level that could cause occasional accidental success?

Reply:
I think of it as being so certain to happen, that I used the word "would" - the number of cycles that it would take for global memory writes by one block to be seen by another block is so large compared to the small number of cycles in the loop that it seems practically impossible for this kernel to ever complete its execution consistently - you'd have to get lucky for every element of every row in one run.

Reply:
> practically impossible for this kernel to ever complete its execution consistently

practically impossible for this kernel to ever complete its execution consistently **if the sync was omitted or didn't work**, that is :-)

Reply:
indeed :) ok, on the basis that it's spectacularly unlucky to hit this otherwise, and the context is about adding the thing to make it work, let's leave it :)